src/realm/object-store/sync/app.cpp

```diff
@@ -723,85 +744,180 @@ std::string App::url_for_path(const std::string& path = "") const
 {
     request.timeout_ms = default_timeout_ms;
 
+    // Normal do_request operation, just send the request to the server and return the response
     if (m_sync_manager->app_metadata()) {
-        m_config.transport->send_request_to_server(std::move(request), std::move(completion));
-        return;
+        m_config.transport->send_request_to_server(
+            std::move(request), [completion = std::move(completion), anchor = shared_from_this()](
+                                    const Request& request, const Response& response) mutable {
+                // If the response contains a redirection, then process it
+                if (response.http_status_code == static_cast<int>(realm::util::HTTPStatus::MovedPermanently)) {
+                    anchor->handle_redirect_response(std::move(const_cast<Request&>(request)), response,
+                                                     std::move(completion));
+                }
+                else {
+                    completion(std::move(response));
+                }
+            });
+        return; // early return
     }
 
-    // if we do not have metadata yet, we need to initialize it and send the
-    // request once that's complete
-    init_app_metadata([completion = std::move(completion), request = std::move(request),
-                       anchor = shared_from_this()](const util::Optional<Response>& error) mutable {
-        if (error) {
-            return completion(std::move(*error));
-        }
+    // if we do not have metadata yet, update the metadata and resend the request
+    update_metadata_and_resend(std::move(request), std::move(completion));
+}
 
-        // if this is the first time we have received app metadata, the
-        // original request will not have the correct URL hostname for
-        // non global deployments.
-        auto app_metadata = anchor->m_sync_manager->app_metadata();
-        if (app_metadata && app_metadata->deployment_model != "GLOBAL" &&
-            request.url.rfind(anchor->m_base_url, 0) != std::string::npos) {
-            request.url.replace(0, anchor->m_base_url.size(), app_metadata->hostname);
+void App::handle_redirect_response(Request&& request, const Response& response,
+                                   UniqueFunction<void(const Response&)>&& completion)
+{
+    // Permanent redirect - get the location and init the metadata again
+    // Look for case insensitive redirect "location" in headers
+    auto location = response.headers.find("location");
+    if (location == response.headers.end() || location->second.empty()) {
+        // Location not found in the response, pass error response up the chain
+        Response error;
+        error.http_status_code = response.http_status_code;
+        error.client_error_code = ClientErrorCode::redirect_error;
+        error.body = "Redirect response missing location header";
+        return completion(std::move(error)); // early return
+    }
+
+    if (request.max_redirects) {
```

Comment:
if max_redirects was just an int that defaults to zero, you could do 
```
if (++request.redirect_count > max_http_redirects) {
   ... handle the error ...
}
```

Reply:
Thanks - I like that better